dbms/src/Flash/Pipeline/Schedule/WaitReactor.cpp

```diff
@@ -0,0 +1,163 @@
+// Copyright 2023 PingCAP, Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <Common/Exception.h>
+#include <Common/setThreadName.h>
+#include <Flash/Pipeline/Schedule/Task/TaskHelper.h>
+#include <Flash/Pipeline/Schedule/TaskScheduler.h>
+#include <Flash/Pipeline/Schedule/WaitReactor.h>
+#include <assert.h>
+#include <common/likely.h>
+#include <common/logger_useful.h>
+#include <errno.h>
+#ifdef __x86_64__
+#include <immintrin.h>
+#endif
+#include <magic_enum.hpp>
+
+namespace DB
+{
+namespace
+{
+class Spinner
+{
+public:
+    explicit Spinner(TaskThreadPool & task_thread_pool_)
+        : task_thread_pool(task_thread_pool_)
+    {}
+
+    // return true if the task is not in waiting status.
+    bool await(TaskPtr && task)
+    {
+        assert(task);
+        TRACE_MEMORY(task);
+        auto status = task->await();
+        switch (status)
+        {
+        case ExecTaskStatus::WAITING:
+            return false;
+        case ExecTaskStatus::RUNNING:
+            ready_tasks.push_back(std::move(task));
+            return true;
+        case FINISH_STATUS:
+            task.reset();
+            return true;
+        default:
+            RUNTIME_ASSERT(false, "Unexpected task state {}", magic_enum::enum_name(status));
+        }
+    }
+
+    void submitAndTryYield()
+    {
+        if (ready_tasks.empty())
+        {
+            tryYield();
+        }
+        else
+        {
+            task_thread_pool.submit(ready_tasks);
+            reset();
+        }
+    }
+
+private:
+    void reset()
+    {
+        ready_tasks.clear();
+        spin_count = 0;
+    }
+
+    void tryYield()
+    {
+        assert(ready_tasks.empty());
+        ++spin_count;
+
+        if (spin_count != 0 && spin_count % 64 == 0)
+        {
+#ifdef __x86_64__
+            _mm_pause();
```

Comment:
Does it consider the intel pause issue: https://aloiskraus.wordpress.com/2018/06/16/why-skylakex-cpus-are-sometimes-50-slower-how-intel-has-broken-existing-code/

Reply:
..no..

Reply:
`_mm_pause` has removed.